src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+            }
+        )
+    }
+}
+
+impl ProgPoint {
+    #[inline(always)]
+    pub fn new(inst: Inst, pos: InstPosition) -> Self {
+        let bits = ((inst.0 as u32) << 1) | (pos as u8 as u32);
+        Self { bits }
+    }
+    #[inline(always)]
+    pub fn before(inst: Inst) -> Self {
+        Self::new(inst, InstPosition::Before)
+    }
+    #[inline(always)]
+    pub fn after(inst: Inst) -> Self {
+        Self::new(inst, InstPosition::After)
+    }
+    #[inline(always)]
+    pub fn inst(self) -> Inst {
+        // Cast to i32 to do an arithmetic right-shift, which will
+        // preserve an `Inst::invalid()` (which is -1, or all-ones).
+        Inst::new(((self.bits as i32) >> 1) as usize)
+    }
+    #[inline(always)]
+    pub fn pos(self) -> InstPosition {
+        match self.bits & 1 {
+            0 => InstPosition::Before,
+            1 => InstPosition::After,
+            _ => unreachable!(),
+        }
+    }
+    #[inline(always)]
+    pub fn next(self) -> ProgPoint {
+        Self {
+            bits: self.bits + 1,
+        }
+    }
+    #[inline(always)]
+    pub fn prev(self) -> ProgPoint {
+        Self {
+            bits: self.bits - 1,
+        }
+    }
+    #[inline(always)]
+    pub fn to_index(self) -> u32 {
+        self.bits
+    }
+    #[inline(always)]
+    pub fn from_index(index: u32) -> Self {
+        Self { bits: index }
+    }
+}
+
+/// An instruction to insert into the program to perform some data movement.
+#[derive(Clone, Debug)]
+pub enum Edit {
+    /// Move one allocation to another. Each allocation may be a
+    /// register or a stack slot (spillslot). However, stack-to-stack
+    /// moves will never be generated.
+    ///
+    /// `to_vreg`, if defined, is useful as metadata: it indicates
+    /// that the moved value is a def of a new vreg.
+    ///
+    /// `Move` edits will be generated even if src and dst allocation
+    /// are the same if the vreg changes; this allows proper metadata
+    /// tracking even when moves are elided.
+    Move {
+        from: Allocation,
+        to: Allocation,
+        to_vreg: Option<VReg>,
+    },
+    /// Define blockparams' locations. Note that this is not typically
+    /// turned into machine code, but can be useful metadata (e.g. for
+    /// the checker).
+    BlockParams {
+        vregs: Vec<VReg>,
+        allocs: Vec<Allocation>,
+    },
+    /// Define a particular Allocation to contain a particular VReg. Useful
+    /// for the checker.
+    DefAlloc { alloc: Allocation, vreg: VReg },
+}
+
+/// A machine envrionment tells the register allocator which registers
+/// are available to allocate and what register may be used as a
+/// scratch register for each class, and some other miscellaneous info
+/// as well.
+#[derive(Clone, Debug)]
+pub struct MachineEnv {
+    /// Physical registers. Every register that might be mentioned in
+    /// any constraint must be listed here, even if it is not
+    /// allocatable under normal conditions.
```

Comment:
Yes, the value should still reach the register correctly. The allocatable-register lists (`preferred` and `non_preferred`) are used only when trying to find a register for an "any register" requirement.

Although, now that I consider this case more: this may mean that the value will live in the non-allocatable register beyond the extent of just the use or def. This is because we won't have any reservations of the register that would otherwise prevent it, and so if we derive the requirement as "must be in rN" where rN is non-allocatable (but requested as fixed), it will appear free and we'll use it.

Thank you for probing these edge-cases more: this is leading me to think that I need to precisely specify in a "how-to use" guide of sorts the expected patterns for lowering e.g. an instruction that requires a value in a fixed but non-allocatable reg. The reason that this isn't as fleshed out as it could be is that the regalloc.rs shim's way of doing things uses pinned vregs for all fixed register mentions, of course. That is probably, in the short term, the best way to handle non-allocatable registers.